src/mongocrypt-kek.c

```diff
@@ -0,0 +1,220 @@
+                             _mongocrypt_kek_t *kek,
+                             mongocrypt_status_t *status)
+{
+   char *kms_provider = NULL;
+   bool ret = false;
+
+   if (!_mongocrypt_parse_required_utf8 (
+          bson, "provider", &kms_provider, status)) {
+      goto done;
+   }
+
+   if (0 == strcmp (kms_provider, "aws")) {
+      kek->kms_provider = MONGOCRYPT_KMS_PROVIDER_AWS;
+      if (!_mongocrypt_parse_required_utf8 (
+             bson, "key", &kek->provider.aws.cmk, status)) {
+         goto done;
+      }
+      if (!_mongocrypt_parse_required_utf8 (
+             bson, "region", &kek->provider.aws.region, status)) {
+         goto done;
+      }
+      if (!_mongocrypt_parse_optional_endpoint (
+             bson, "endpoint", &kek->provider.aws.endpoint, status)) {
+         goto done;
+      }
+   } else if (0 == strcmp (kms_provider, "local")) {
+      kek->kms_provider = MONGOCRYPT_KMS_PROVIDER_LOCAL;
+   } else if (0 == strcmp (kms_provider, "azure")) {
+      kek->kms_provider = MONGOCRYPT_KMS_PROVIDER_AZURE;
+      if (!_mongocrypt_parse_required_endpoint (
+             bson,
+             "keyVaultEndpoint",
+             &kek->provider.azure.key_vault_endpoint,
+             status)) {
+         goto done;
+      }
+
+      if (!_mongocrypt_parse_required_utf8 (
+             bson, "keyName", &kek->provider.azure.key_name, status)) {
+         goto done;
+      }
+
+      if (!_mongocrypt_parse_optional_utf8 (
+             bson, "keyVersion", &kek->provider.azure.key_version, status)) {
+         goto done;
+      }
+   } else if (0 == strcmp (kms_provider, "gcp")) {
+      kek->kms_provider = MONGOCRYPT_KMS_PROVIDER_GCP;
+      if (!_mongocrypt_parse_optional_endpoint (
+             bson, "endpoint", &kek->provider.gcp.endpoint, status)) {
+         goto done;
+      }
+
+      if (!_mongocrypt_parse_required_utf8 (
+             bson, "projectId", &kek->provider.gcp.project_id, status)) {
+         goto done;
+      }
+
+      if (!_mongocrypt_parse_required_utf8 (
+             bson, "location", &kek->provider.gcp.location, status)) {
+         goto done;
+      }
+
+      if (!_mongocrypt_parse_required_utf8 (
+             bson, "keyRing", &kek->provider.gcp.key_ring, status)) {
+         goto done;
+      }
+
+      if (!_mongocrypt_parse_required_utf8 (
+             bson, "keyName", &kek->provider.gcp.key_name, status)) {
+         goto done;
+      }
+
+      if (!_mongocrypt_parse_optional_utf8 (
+             bson, "keyVersion", &kek->provider.gcp.key_version, status)) {
+         goto done;
+      }
+   } else {
+      CLIENT_ERR ("unrecognized KMS provider: %s", kms_provider);
+      goto done;
+   }
+
+   ret = true;
+done:
+   bson_free (kms_provider);
+   return ret;
+}
+
+bool
+_mongocrypt_kek_append (const _mongocrypt_kek_t *kek,
+                        bson_t *bson,
+                        mongocrypt_status_t *status)
+{
+   if (kek->kms_provider == MONGOCRYPT_KMS_PROVIDER_AWS) {
```

Comment:
Add an else if (kms_provider != 0) that asserts or aborts, your preference. Since the monocrypt-kek-private.h header talks about kms_provider being a bitset, I want to make sure there is no accidental misuse where more then 1 bit gets set in kms_provider.

Reply:
Good thinking, done.